### PR TITLE
Don't allocate a mergeSeries if there is only one series to merge.

### LINKED ISSUE #3715: CPU Usage triples after upgrade to v2.1.0
**What did you do?**

Upgraded Prometheus from v2.0.0 to v2.1.0 (using the official Docker image)

**What did you expect to see?**

Similar or better performance figures.

**What did you see instead? Under which circumstances?**

CPU Usage almost tripled and rule evaluation times increased, we observed the same on two different deployment (two separate K8s clusters), after failing back to v2.0.0 figures are back to normal.

Cluster 1 [CPU]:
image

Cluster 2 [CPU]:
image

Cluster 1 [Rule evluation duration]:
image

Cluster 2 [Rule evluation duration]: graph looks about the same, but this cluster is smaller:
image

Except changing the docker image version, everything remained unchanged.

**Environment**

* System information:

`Linux 4.12.10-coreos x86_64` (Runs in K8s 1.5.3 Cluster)

* Prometheus version:

```
prometheus, version 2.1.0 (branch: HEAD, revision: 85f23d82a045d103ea7f3c89a91fba4a93e6367a)
  build user:       root@6e784304d3ff
  build date:       20180119-12:01:23
  go version:       go1.9.2
```

* Alertmanager version:

	N/A

* Prometheus configuration file:
```
Can provide if needed
```

* Alertmanager configuration file:
```
N/A
```

* Logs:
```
Can provide if needed
```

### PR DESCRIPTION
**Pull request by tomwilkie**
~Might help #3715~

Doesn't actually seem to help, but are good to have none-the-less.

### CODE REVIEW

**Comment by tomwilkie** [comment]
Circle CI failures are unrelated (not sure whats going on, but theres nothing wrong with this PR)

**Comment by grobie** [comment]
Do we have a benchmark showing the impact in terms of allocations? If not, could we add one, so that we can catch regressions in future releases?

**Comment by tomwilkie** [comment]
> Do we have a benchmark showing the impact in terms of allocations?

Not thats I know of :-(

> If not, could we add one, so that we can catch regressions in future releases?

Sure, will try and get it done today.

**Comment by tomwilkie** [comment]
Before these changes:
```
$ go test -bench=MergeSeriesSet ./storage/
BenchmarkNoMergeSeriesSet_100_100-4      	500000000	         2.78 ns/op
BenchmarkMergeSeriesSet_1_100_100-4      	200000000	         6.88 ns/op
```

After:
```
$ go test -bench=MergeSeriesSet ./storage/
BenchmarkNoMergeSeriesSet_100_100-4      	1000000000	         2.51 ns/op
BenchmarkMergeSeriesSet_1_100_100-4      	1000000000	         2.89 ns/op
```

Basically, calling MergeSeriesSet on a single SeriesSet is now no extra cost.  I added some extra benchmarks but they need more reasoning first.

File: storage/fanout_test.go [hunk 1]
```diff
@@ -218,13 +219,58 @@ func (m *mockSeriesSet) Err() error {
...
+	benchmarkDrain(seriesSet, b)
+}
+
+func BenchmarkMergeSeriesSet_100_100_100(b *testing.B) {
+	seriesSet := makeMergeSeriesSet(100, 100, 100)
+	benchmarkDrain(seriesSet, b)
+}
+
+func BenchmarkMergeSeriesSet_1000_100_100(b *testing.B) {
+	seriesSet := makeMergeSeriesSet(100, 100, 100)
```
Comment (grobie) [inline storage/fanout_test.go]: This is identical to the one above.

File: storage/fanout_test.go [hunk 2]
```diff
@@ -218,13 +219,58 @@ func (m *mockSeriesSet) Err() error {
...
+		}
 	}
 }
+
+func BenchmarkNoMergeSeriesSet_100_100(b *testing.B) {
+	seriesSet := makeSeriesSet(100, 100)
+	benchmarkDrain(seriesSet, b)
+}
+
+func BenchmarkMergeSeriesSet_1_100_100(b *testing.B) {
```
Comment (grobie) [inline storage/fanout_test.go]: Why not using sub-benchmarks?

Comment (tomwilkie) [reply storage/fanout_test.go]: Didn't know about those!  Much nicer.

File: storage/fanout_test.go (hunk 1 above)
Comment (tomwilkie) [reply storage/fanout_test.go]: Roger; I removed it.

**Comment by tomwilkie** [comment]
@grobie can I take the 'Looks great' as an LGTM :-) ?

**Pull request merged**